packages/samples/headless-commerce-ssr-remix/app/components/add-to-cart-button.tsx

```diff
@@ -0,0 +1,51 @@
+import {ExternalCartItem} from '@/client/external-cart-api';
+import {useCart} from '@/lib/commerce-engine';
+import {useFetcher} from '@remix-run/react';
+import {useEffect} from 'react';
+
+type AddToCartButtonProps = {
+  productId: string;
+  price: number;
+  name: string;
+};
+
+export default function AddToCartButton({
+  productId,
+  price,
+  name,
+}: AddToCartButtonProps) {
+  const fetcher = useFetcher<ExternalCartItem>();
+  const {methods} = useCart();
+
+  useEffect(() => {
```

Comment:
The idea here is to ensure that we emit the `ec.cartAction` event _after_ the external cart API call has returned.

In general, I believe this is desirable, because this way if an external API call fails, we won't end up with bad data / an out-of-sync cart state in the Headless commerce engine.

Here, I'm simulating an API that would return the updated item data when an item is added in the cart, and showing that it's a good idea to use this data when calling the cart controller's `updateItemQuantity` method.

Reply:
![die-fetcher](https://github.com/user-attachments/assets/bf4fb868-0090-455b-83d6-b33f48e1de36)